README.md
# lambda-resize-image

[![Build Status](https://travis-ci.org/apoca/lambda-resize-image.svg?branch=master)](https://travis-ci.org/apoca/lambda-resize-image)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/174785b0f3a249e2a2f8482542e8f557)](https://app.codacy.com/app/apoca/lambda-resize-image?utm_source=github.com&utm_medium=referral&utm_content=apoca/lambda-resize-image&utm_campaign=Badge_Grade_Dashboard)
[![Maintainability](https://api.codeclimate.com/v1/badges/c0bcf34c657a91f453e6/maintainability)](https://codeclimate.com/github/apoca/lambda-resize-image/maintainability)
[![npm version](https://badge.fury.io/js/lambda-resize-image.svg)](https://badge.fury.io/js/lambda-resize-image)
[![dependencies Status](https://david-dm.org/apoca/lambda-resize-image/status.svg)](https://david-dm.org/apoca/lambda-resize-image)
[![devDependencies Status](https://david-dm.org/apoca/lambda-resize-image/dev-status.svg)](https://david-dm.org/apoca/lambda-resize-image?type=dev)
[![npm](https://img.shields.io/npm/l/lambda-images-resizer.svg)]()

An AWS Lambda Function to resize images automatically with API Gateway and S3 for imagemagick tasks. When an image is called on AWS Api Gateway, this package will resize it and send it to the S3.

## Contents

- [lambda-resize-image](#lambda-resize-image)
  - [Contents](#contents)
  - [Requirements](#requirements)
  - [What is it?](#what-is-it)
  - [Description](#description)
  - [Features](#features)
  - [Instalation](#instalation)
    - [AWS credentials](#aws-credentials)
    - [Deploy to Amazon AWS](#deploy-to-amazon-aws)
  - [Usage (image restrictions resize)](#usage-image-restrictions-resize)
    - [Example URL usage](#example-url-usage)
  - [Environment variables](#environment-variables)
    - [Environment variables for serverless](#environment-variables-for-serverless)
  - [Local development](#local-development)
    - [Example request](#example-request)
      - [Configuration Parameters](#configuration-parameters)
  - [Feedback](#feedback)
  - [Contributing](#contributing)
  - [License](#license)
  - [Author](#author)

## Requirements

- Node.js - AWS Lambda supports versions of **10.20.1** or above (Recommended: **12.X**).

## New version 2.0

We've changed our approache to remove redirects from our responses. Then, we are returning a binary file base 64 image file.

### Steps to version 2.0

You must have AWS CLI installed to execute a command from your console:

`aws apigateway update-integration-response --rest-api-id <API_ID> --resource-id <RESOURCE_ID> --http-method GET  --status-code 200  --patch-operations '[{"op" : "replace", "path" : "/contentHandling", "value" : "CONVERT_TO_BINARY"}]'`

In API GW -> Settings -> Binary Media Types and add:

`*/*`

## What is it?

It's AWS Lambda, which is a compute service that lets you run code without provisioning or managing servers.

> _[Read more about AWS Lambda.](http://docs.aws.amazon.com/lambda/latest/dg/welcome.html)_

## Description

The combination of API Gateway and Lambda is very powerful. It allows you to build some complex functionalities without maintaining any virtual machines yourself. Lambda can be hooked up to many other (AWS) Services including S3. That's why I decided to build an AWS Lambda Function to resize images automatically with API Gateway and S3 for imagemagick tasks. When an image is called on AWS S3 bucket (via API Gateway), this package will resize it and send it to the S3 before redirecting you to the new path of the image (aws bucket url or CDN).

## Features

- Use [Serverless Framework](https://github.com/serverless/serverless#features)
- Use [Serverless Webpack](https://github.com/serverless-heaven/serverless-webpack)
- Use [Serverless Offline](https://github.com/dherault/serverless-offline)
- The image conversion endpoint by API Gateway or cloudfront URL.

## Instalation

First, add Serverless globally:

`npm install -g serverless`

Then, clone the repository into your local environment:

```bash
git clone https://github.com/apoca/lambda-resize-image
cd lambda-resize-image
npm install
```

Modify and set the .env.example to env.yml with your data. (You are supposed to have already configured your [local variables](#environment-variables)).

### AWS credentials

To run [local development](#local-development) you also might need to configure you [aws credentials](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/installing-jssdk.html), or you can set them to what I've shown below.

### Deploy to Amazon AWS

You can also check if you have everything installed in the correct way:

`$ serverless`

To deploy from your environment to Amazon AWS, you must:

`$ serverless deploy --stage dev` or `$ serverless deploy --stage prod` for production configurations.

Then, serverless will package, validate and upload your stack to Amazon AWS. It will probably look like this:

```bash
Serverless: Packaging service...
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service .zip file to S3 (1.52 KB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
..............
Serverless: Stack update finished...
Service Information
service: resizeS3Image
stage: development
region: eu-west-1
stack: resizeS3Image-development
api keys:
  None
endpoints:
  GET - https://<api_key_here>.execute-api.eu-west-1.amazonaws.com/dev/{key+}
functions:
  handler: resize-dev-image
Serverless: Removing old service artifacts from S3...
```

When upload to AWS Lambda, the project will bundle only needed files - no dev dependencies will be included.

## Usage (image restrictions resize)

To restrict the dimensions, we put a `const var` in the `handle.js`:

```javascript
const ALLOWED_DIMENSIONS = {
  width: 1800,
  height: 1800
};
```

You can also change the url endpoint `https://<api_key_here>.execute-api.eu-west-1.amazonaws.com/dev/{key+}` to another one more tiny and cachable (cloudfront), you can also configure in you Api Gateway (lambda service) a **Custom Domain Name**.

### Example URL usage

`https://<api_key_here>.execute-api.eu-west-1.amazonaws.com/dev/<KEY_S3_IMAGE_HERE>?width=<WIDTH>&height=<HEIGHT>`

or with you own Custom Domain Names:

`https://api.yourdomain.com/<KEY_S3_IMAGE_HERE>?width=<WIDTH>&height=<HEIGHT>`

![Example usage of lambda resize image](https://s3-eu-west-1.amazonaws.com/lambda-resize-image/example-usage-lambda-resize-image.png)

## Environment variables

- URL - AWS URL S3 bucket or your CDN url to the BUCKET. (not required in version 2.0)
- BUCKET - AWS S3 bucket. (required)

### Environment variables for serverless

- AWS_ACCESS_KEY_ID - AWS key
- AWS_SECRET_ACCESS_KEY - AWS Secret

## Local development

1. `$ npm i -g serverless`
2. `$ npm i`
3. `$ brew install imagemagick`
4. `$ serverless offline start`

Note that you will need to be into the root repository. The last command (4.) will spin up an [serverless-offline](https://github.com/dherault/serverless-offline) version of an API Gateway, that will simulate the real one. Once it is running, you can see all the requests on your command line.

### Example request

`http://localhost:3000<YOUR_KEYNAME_TO_IMAGE>?width=<WIDTH>&height=<HEIGHT>`

#### Configuration Parameters

| Parameter |  type   | description                                        |
| :-------: | :-----: | -------------------------------------------------- |
|   width   | Integer | Required. Will resize image via imagemagic resize. |
|  height   | Integer | Optional. Will resize image via imagemagic crop.   |

## Feedback

We'd love to get feedback on how you're using lambda-resize-image and things we could add to make this tool better. Feel free to contact us at vieira@miguelvieira.com.pt

## Contributing

If you'd like to contribute to the project, feel free to submit a PR. See more: [CODE_OF_CONDUCT.md](CODE_OF_CONDUCT.md)

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Author

- **Miguel Vieira** - _Initial work_ - [apoca](https://github.com/apoca)

See also the list of [contributors](https://github.com/apoca/lambda-resize-image/contributors) who participated in this project.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.2.0] - 2020-05-09

    ### Changed

    - Improved image cache-control to 1y

## [2.1.0] - 2020-05-08

    ### Changed

    - Improved image quality returned

## [2.0.0] - 2020-04-30

    ### Changed

    - Fixed a lot of vulnerabilities
    - Updated handler.js to return binary images not redirects
    - Updated webpack.config.js to better "bundle" the package to Lambda AWS amazon
    - Improved serverless.yml

## [1.3.3] - 2019-04-24

    ### Changed

    - Fixed 2 vulnerabilities
    - Updated puObject IAM AWS to putObjectAcl

## [1.3.2] - 2018-10-13

     ### Changed

     - Updated fs unlink callback function

## [1.3.1] - 2018-10-13

    ### Changed

    - Fixed 12 vulnerabilities

## [1.3.0] - 2018-09-12

    ### Added

    - Add missing yarn.lock
    - Add CHANGELOG.md

    ### Changed

    - Improving information about the image output
    - Added some other new keywords and improved readme file.
    - Update image module
    - Update handler module
    - Update utils module
    - Update version package.json
    - Improved cors serverless api gateway

## [1.2.2] - 2018-09-10

    ### Changed

    - Improved some unit test
    - Improved serverless configuration

    ### Removed

    - Removed unnecessary API_URL envirenment variable

## [1.2.1] - 2018-09-09

    ### Added

    - Created a new environment variable for API_URL into serverless (env.yml)
    - Created a new function getImageKey

    ### Changed

    - Improved the text of API_URL in README.md
    - Improved the contributors obkect inside package.json

## [1.2.0] - 2018-09-04

    ### Added

    - Starting develop unit test with jest

[unreleased]: https://github.com/apoca/lambda-resize-image/compare/v2.2.0...HEAD
[2.2.0]: https://github.com/apoca/lambda-resize-image/compare/v2.1.0..v2.2.0
[2.1.0]: https://github.com/apoca/lambda-resize-image/compare/v2.0.1..v2.1.0
[2.0.1]: https://github.com/apoca/lambda-resize-image/compare/v2.0.0..v2.0.1
[2.0.0]: https://github.com/apoca/lambda-resize-image/compare/v1.3.3..v2.0.0
[1.3.3]: https://github.com/apoca/lambda-resize-image/compare/v1.3.2..v1.3.3
[1.3.2]: https://github.com/apoca/lambda-resize-image/compare/v1.3.1..v1.3.2
[1.3.1]: https://github.com/apoca/lambda-resize-image/compare/v1.3.0..v1.3.1
[1.3.0]: https://github.com/apoca/lambda-resize-image/compare/v1.2.2..v1.3.0
[1.2.2]: https://github.com/apoca/lambda-resize-image/compare/v1.2.1...v1.2.2
[1.2.1]: https://github.com/apoca/lambda-resize-image/compare/v1.2.0...v1.2.1
[1.2.0]: https://github.com/apoca/lambda-resize-image/compare/v1.1.0...v1.2.0


test/globals.js
//process.env.BUCKET = 'Foo';
//process.env.URL = 'Bar';


tests/handler_test.js
//import your handler file or main file of Lambda
const { handler } = require('../src/index');

//Call your exports function with required params
//In AWS lambda these are event, content, and callback
//event and content are JSON object and callback is a function
//In my example i'm using empty JSON
handler(
  {
    queryStringParameters: {
      width: 180,
      height: null
    },
    path:
      'https://esolidar-proto-uploads.s3-eu-west-1.amazonaws.com/brands/580ded66-0364-40f3-8f37-5451c0a8010a.jpg'
  }, //event
  {}, //content
  function(data, ss) {
    //callback function with two arguments
    //console.log(data, ss);
  }
);


test/config/globals.js
//process.env.BUCKET = 'my-bucket-here';
//process.env.URL = 'https://localhost:3000';


src/index.js
const { resizeImage, getImage } = require('./lib/image');
const url = require('url');
const { generateS3Key } = require('./lib/utils');
const { BUCKET, URL } = process.env;

module.exports.imageprocess = event =>
  new Promise((resolve, reject) => {
    const queryParameters = event.queryStringParameters || {};
    const path = event.path;
    const imageKey = url.parse(path).pathname.replace(/^\//g, '');

    if (!BUCKET || !URL) {
      return reject('Error: Set environment variables BUCKET and URL.');
    }

    const size = {
      width:
        queryParameters.width === 'AUTO'
          ? null
          : parseInt(queryParameters.width),
      height:
        queryParameters.height === 'AUTO'
          ? null
          : parseInt(queryParameters.height)
    };

    if (!queryParameters.width && !queryParameters.height) {
      return getImage(imageKey)
        .then(resolve)
        .catch(reject);
    } else {
      return getImage(generateS3Key(imageKey, size)).then(resolve);
    }

    if (!size.width) {
      return reject('The parameter width is a required field to resize.');
    }

    return resizeImage(imageKey, size)
      .then(resolve)
      .catch(reject);
  });


src/lib/image.js
/**
 * Module dependencies.
 */

import { S3 as _S3 } from 'aws-sdk';
import { crop, resize } from 'imagemagick';
import { tmpDir } from 'os';
import { resizeCallback, generateS3Key } from './utils';

/**
 * Export `getImage` util.
 */

export function getImage(key) {
  return new Promise((resolve, reject) => {
    const S3 = new _S3({
      signatureVersion: 'v4',
    });

    S3.getObject(
      {
        Bucket: process.env.BUCKET,
        Key: key,
      },
      (err) => {
        if (err) return reject(err);

        resolve({
          statusCode: 301,
          headers: {
            Location: `${process.env.URL}/${key}`,
          },
        });
      }
    );
  });
}

/**
 * Export `checkKeyExists` util.
 */

export function checkKeyExists(key, size) {
  return new Promise((resolve, reject) => {
    const S3 = new _S3({
      signatureVersion: 'v4',
    });

    S3.headObject(
      {
        Bucket: process.env.BUCKET,
        Key: generateS3Key(key, size),
      },
      (err) => {
        if (err && err.code === 'NotFound')
          return resizeImage(key, size).then(resolve).catch(reject);

        resolve({
          statusCode: 301,
          headers: {
            Location: `${process.env.URL}/${generateS3Key(key, size)}`,
          },
        });
      }
    );
  });
}

/**
 * Export `resizeImage` util.
 */

export function resizeImage(key, size) {
  return new Promise((resolve, reject) => {
    const S3 = new _S3({
      signatureVersion: 'v4',
    });

    S3.getObject(
      {
        Bucket: process.env.BUCKET,
        Key: key,
      },
      (err, data) => {
        if (err) return reject(err);

        if (size.width && size.height) {
          const tmpImageName = `${tmpDir}/resized.${process.env.BUCKET}.${size.width}.${size.height}`;

          crop(
            {
              width: size.width,
              height: size.height,
              srcData: data.Body,
              dstPath: tmpImageName,
              quality: 1,
              gravity: 'Center',
            },
            (err) => {
              if (err) return reject(err);

              resolve(
                resizeCallback(
                  err,
                  data.ContentType,
                  generateS3Key(key, size),
                  tmpImageName
                )
              );
            }
          );
        } else if (size.width) {
          const tmpImageName = `${tmpDir}/resized.${process.env.BUCKET}.${size.width}`;

          resize(
            {
              width: size.width,
              srcData: data.Body,
              dstPath: tmpImageName,
            },
            (err) => {
              if (err) return reject(err);

              resolve(
                resizeCallback(
                  err,
                  data.ContentType,
                  generateS3Key(key, size),
                  tmpImageName
                )
              );
            }
          );
        } else {
          resolve({
            statusCode: 301,
            headers: {
              Location: `${process.env.URL}/${generateS3Key(key, size)}`,
            },
          });
        }
      }
    );
  });
}


src/lib/utils.js
/**
 * Module dependencies.
 */

import { S3 as _S3 } from 'aws-sdk';
import { unlink, readFileSync } from 'fs';
const PathReg = new RegExp('(.*)/(.*)');

/**
 * Export `resizeCallback` util.
 */

export function resizeCallback(error, contentType, newKey, tmpImageName) {
  return new Promise((resolve, reject) => {
    if (error) {
      reject(error);
    } else {
      const S3 = new _S3({
        signatureVersion: 'v4',
      });

      S3.putObject(
        {
          ACL: 'public-read',
          Bucket: process.env.BUCKET,
          Body: readFileSync(tmpImageName),
          ContentType: contentType,
          Key: newKey,
        },
        (err) => {
          if (err) return reject(err);

          unlink(tmpImageName, (err) => {
            if (err) throw err;
          });

          resolve({
            statusCode: 301,
            headers: {
              Location: `${process.env.URL}/${newKey}`,
            },
          });
        }
      );
    }
  });
}

/**
 * Export `generateS3Key` util.
 */

export function generateS3Key(key, size) {
  let parts = PathReg.exec(key);

  if (!parts) {
    return key;
  }

  let oldKey = parts[1];
  let filename = parts[2];
  let width = size.width || null;
  let height = size.height || 'AUTO';

  if (width) {
    return `${oldKey}/${width}x${height}/${filename}`;
  }

  return `${oldKey}/${filename}`;
}


webpack.config.js
const slsw = require('serverless-webpack');
const path = require('path');
const nodeExternals = require('webpack-node-externals');

module.exports = {
  // Specify the entry point for our app.
  entry: slsw.lib.entries,
  target: 'node',
  // Generate sourcemaps for proper error messages
  devtool: 'source-map',
  externals: [
    nodeExternals({
      modulesFromFile: true,
    }),
  ],
  mode: slsw.lib.webpack.isLocal ? 'development' : 'production',
  optimization: {
    // We do not want to minimize our code.
    minimize: false,
  },
  performance: {
    // Turn off size warnings for entry points
    hints: false,
  },
  module: {
    // Run babel on all .js files and skip those in node_modules
    rules: [
      {
        test: /\.js$/,
        loader: 'babel-loader',
        include: __dirname,
        exclude: /node_modules/,
      },
    ],
  },
  // Specify the output file containing our bundled code
  output: {
    libraryTarget: 'commonjs',
    path: path.resolve(__dirname, '.webpack'),
    filename: '[name].js',
  },
};


src/lib/amazon-s3-uri.js
'use strict';

const url = require('url');
const { REGION } = process.env;

const ENDPOINT_PATTERN = /^(.+\.)?s3[.-]([a-z0-9-]+)\./;
const DEFAULT_REGION = REGION || 'eu-west-1';

// TODO: support versionId
// TODO: encode uri before testing
// TODO: support dualstack http://docs.aws.amazon.com/AmazonS3/latest/dev/dual-stack-endpoints.html

/**
 * A URI wrapper that can parse out information about an S3 URI
 *
 * Directly adapted from https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/AmazonS3URI.java
 *
 * @param {String} uri - the URI to parse
 * @throws {TypeError|Error}
 */
function AmazonS3URI(uri) {
  if (!new.target) {
    return new AmazonS3URI(uri);
  }
  this.uri = url.parse(uri);

  if (this.uri.protocol === 's3:') {
    this.region = DEFAULT_REGION;
    this.isPathStyle = false;
    this.bucket = this.uri.host;

    if (!this.bucket) {
      throw new Error(`Invalid S3 URI: no bucket: ${uri}`);
    }

    if (!this.uri.path || this.uri.path.length <= 1) {
      // s3://bucket or s3://bucket/
      this.key = null;
    } else {
      // s3://bucket/key
      // Remove the leading '/'.
      this.key = this.uri.path.substring(1);
    }
    if (this.key !== null) {
      this.key = decodeURIComponent(this.key);
    }
    return;
  }

  if (!this.uri.host) {
    throw new Error(`Invalid S3 URI: no hostname: ${uri}`);
  }

  const matches = this.uri.host.match(ENDPOINT_PATTERN);
  if (!matches) {
    throw new Error(
      `Invalid S3 URI: hostname does not appear to be a valid S3 endpoint: ${uri}`
    );
  }

  const prefix = matches[1];
  if (!prefix) {
    // No bucket name in the host; parse it from the path.
    this.isPathStyle = true;

    if (this.uri.path === '/') {
      this.bucket = null;
      this.key = null;
    } else {
      const index = this.uri.path.indexOf('/', 1);
      if (index === -1) {
        // https://s3.amazonaws.com/bucket
        this.bucket = this.uri.path.substring(1);
        this.key = null;
      } else if (index === this.uri.path.length - 1) {
        // https://s3.amazonaws.com/bucket/
        this.bucket = this.uri.path.substring(1, index);
        this.key = null;
      } else {
        // https://s3.amazonaws.com/bucket/key
        this.bucket = this.uri.path.substring(1, index);
        this.key = this.uri.path.substring(index + 1);
      }
    }
  } else {
    // Bucket name was found in the host; path is the key.
    this.isPathStyle = false;

    // Remove the trailing '.' from the prefix to get the bucket.
    this.bucket = prefix.substring(0, prefix.length - 1);

    if (!this.uri.path || this.uri.path === '/') {
      this.key = null;
    } else {
      // Remove the leading '/'.
      this.key = this.uri.path.substring(1);
    }
  }

  if (matches[2] !== 'amazonaws') {
    this.region = matches[2];
  } else {
    this.region = REGION;
  }

  if (this.key !== null) {
    this.key = decodeURIComponent(this.key);
  }
}

AmazonS3URI.prototype.DEFAULT_REGION = REGION;

module.exports = AmazonS3URI;


src/handler.js
/**
 * Module dependencies.
 */

import { S3 as _S3 } from 'aws-sdk';
import { generateS3Key } from './lib/utils';
import { isNullOrUndefined } from 'util';
import sharp from 'sharp';

const S3 = new _S3({
  signatureVersion: 'v4',
});

const ALLOWED_DIMENSIONS = {
  width: 1800,
  height: 1800,
};

/**
 * Export `imageprocess` module.
 */

export function imageprocess(event, context, callback) {
  const queryParameters = event.queryStringParameters || {};
  const imageKey = decodeURIComponent(event.pathParameters.key);

  if (!process.env.BUCKET) {
    return callback(null, {
      statusCode: 404,
      body: 'Error: Set environment variables BUCKET.',
    });
  }

  const size = {
    width: isNullOrUndefined(queryParameters.width)
      ? null
      : parseInt(queryParameters.width),
    height: isNullOrUndefined(queryParameters.height)
      ? null
      : parseInt(queryParameters.height),
  };

  if (
    size.width > ALLOWED_DIMENSIONS.width ||
    size.height > ALLOWED_DIMENSIONS.height
  ) {
    return callback(null, {
      statusCode: 403,
      body: 'Error: Image size not permited.',
    });
  }

  if (imageKey) {
    if (!size.width && !size.height) {
      S3.getObject({
        Bucket: process.env.BUCKET,
        Key: imageKey,
      })
        .promise()
        .then((data) =>
          sharp(data.Body).jpeg({ quality: 95, progressive: true }).toBuffer()
        )
        .then((buffer) => {
          // generate a binary response with resized image
          const response = {
            statusCode: 200,
            headers: {
              'Content-Type': 'image/jpeg',
              'Cache-Control': 'public, max-age=31536000',
            },
            body: buffer.toString('base64'),
            isBase64Encoded: true,
          };
          return callback(null, response);
        })
        .catch((err) =>
          callback(null, {
            statusCode: err.statusCode || 404,
            body: JSON.stringify(err),
          })
        );
    } else {
      S3.getObject({
        Bucket: process.env.BUCKET,
        Key: imageKey,
      })
        .promise()
        .then((data) =>
          sharp(data.Body)
            .resize(size.width, size.height)
            .jpeg({ quality: 95, progressive: true })
            .toBuffer()
        )
        .then((buffer) => {
          S3.putObject({
            Body: buffer,
            Bucket: process.env.BUCKET,
            ContentType: 'image/jpeg',
            CacheControl: 'max-age=31536000',
            Key: generateS3Key(imageKey, size),
            ACL: 'public-read',
          })
            .promise()
            .catch((err) => {
              return callback(null, {
                statusCode: err.statusCode || 404,
                body: JSON.stringify(err),
              });
            });
          // generate a binary response with resized image
          const response = {
            statusCode: 200,
            headers: {
              'Content-Type': 'image/jpeg',
              'Cache-Control': 'public, max-age=31536000',
            },
            body: buffer.toString('base64'),
            isBase64Encoded: true,
          };
          return callback(null, response);
        })
        .catch((err) =>
          callback(null, {
            statusCode: err.statusCode || 404,
            body: JSON.stringify(err),
          })
        );
    }
  } else {
    return callback(null, {
      statusCode: 404,
      body: 'Error: Image not found.',
    });
  }
}
